### PR TITLE
Making Scheduler's job related APIs work on the right server as jobs …

### PR DESCRIPTION
**Pull request by suresh-thelkar**
…are sharded in Multi-Server

#### Describe Bug or Feature

In a Multi-Server world as we are sharding jobs, Scheduler should run the jobs in the right server from which they are originated. In order to do this Scheduler should act on the right sd/socket of the server for which the jobs are submitted. In other words Scheduler should call its APIs like run_job, send_run_job, update_job_attr etc on the right sever's sd i.e. it should not use cluster sd for this purpose. 

**Design Document**: https://openpbs.atlassian.net/wiki/spaces/PD/pages/2233991174/Introducing+an+attribute+server+instance+id+for+sharded+objects
**Forum Discussion**:  https://community.openpbs.org/t/adding-a-new-member-to-batch-status-structure/2353/41

For more details on Multi-Server feature please refer to the following.

https://openpbs.atlassian.net/wiki/spaces/PD/pages/1507164161/PBS+Pro+multi-server+project+architecture+and+design

#### Describe Your Change

- Populate Scheduler's internal data structures like job/node objects with the right server sd. So when scheduler accesses the job object it can directly get its server sd easily. 
  The population of the server sds to the job/node objects is done inside IFL layer i.e. while aggregating jobs/nodes itself an extra memeber by the name "server_sock" is added to the batch status. This way by the time Scheduler queries jobs/nodes this info is handy. Scheduler upon querying jobs/nodes it stores "server_sock" in its internal data structures i.e. job_info/node_info objects contain server_sock now.
  
    For example we can make use of this info while calling run_job API inside scheudler
    pbsrc = run_job(resresv->job->server_sock, rr, execvnode, sinfo->has_runjob_hook, err);

-  Scheduler's APIs like run_job, send_attr_updates, pbs_sigjob etc need to act on the right server sd. This PR concentrates on most of the job related APIs like these to act on right server sd so that we can actually see multiple servers functioning in action. There are a few more APIs like pbs_defschreply and handling qrun requests which might also need to act upon right server sd which will be taken up in the next PR. The idea here is other multi-server features like "Scheduler preferring local nodes first, move jobs from a source server to destination server in case if the source server do not have nodes" can benefit out of this PR and they can be worked on immediately.

- Please also note that pbs_asyalterjob is not made multi-server compatible as part of this PR. This is because we have to have special treatment for this like pbs_preempt_jobs() IFL when calling from Scheduler which we are taking care of as part of upcoming Multi-Server PRs. This does not cause any problem as of now for the following reasons.

   - No impact on server functionality
   - As we are making all IFLs work in case of Multi-Server even when they are called from a non scheduler like client/s. So when an IFL is called on virtual fd, the API contacts server instances in some random fashion and get to know the right server for which it sends the request. These changes are already happening as part of the PR #2086

#### Attach Test and Valgrind Logs/Output

Attached are the detailed test logs for Multi-Server.
[SchedJobRelatedAPIS.txt.log](https://github.com/openpbs/openpbs/files/5500242/SchedJobRelatedAPIS.txt.log)

Following are the regression results.
Description: Tests from given sources on platforms UBUNTU1804, SUSE15, UBUNTU1604, CENTOS8, RHEL8, CENTOS7, RHEL7, SLES12, SLES15, WIN2K16
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|4692|3845|2|0|0|0|3843|

The test cases failed are known issues and happening in master also.

### CODE REVIEW

**Comment by suresh-thelkar** [comment]
I am in the process of running valgrind on these code changes. Meanwhile requesting the reviews for this PR.

File: src/lib/Libcmds/batch_status.c
Comment (agrawalravi90) [inline src/lib/Libcmds/batch_status.c:139]: i think we should follow the name scheme in this file and name this 'bs_init' instead

File: src/lib/Libcmds/batch_status.c
Comment (agrawalravi90) [inline src/lib/Libcmds/batch_status.c:141]: check for NULL first?

File: src/scheduler/data_types.h
Comment (agrawalravi90) [inline src/scheduler/data_types.h:576]: can we add it to resource_resv instead? one less indirection to go through when accessing it, and we might need it for reservations as well.

File: src/scheduler/pbs_sched.cpp
Comment (agrawalravi90) [inline src/scheduler/pbs_sched.cpp:573]: Move this check before the for loop

File: src/scheduler/pbs_sched.cpp
Comment (agrawalravi90) [inline src/scheduler/pbs_sched.cpp:577]: i think there's a bug here, you are checking svr_conns_primary.state twice (line 577 and line 579), one of them should be checking svr_conns_secondary.state right?

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]

**Comment by bhroam** [comment]
I see two things here which are major changes to our IFL interface with no design.  First, you are changing the batch_status.  This is the main structure we pass PBS objects in.  Second, you are changing how the different IFL interfaces use the pbs sockets.  Before you do a pbs_connect() and then use *ANY* IFL interface on that socket.  Now that returns a multi-server socket that can only be used for certain calls.  Now you need to use that multi-server interface to do a stat, and then find the socket you need to use for other calls like alterjob or runjob.  Please write up a design for these changes so they can be discussed on the forums.

**Comment by agrawalravi90** [comment]
> Before you do a pbs_connect() and then use ANY IFL interface on that socket. Now that returns a multi-server socket that can only be used for certain calls. Now you need to use that multi-server interface to do a stat, and then find the socket you need to use for other calls like alterjob or runjob. Please write up a design for these changes so they can be discussed on the forums.

Suresh can explain more, but just to add some clarification, clients can still use the fd returned by pbs_connect directly  without having to care about the fd inside each batch_status .. it's just that if they use the fd that comes within the batch_status, then IFL will not bother trying to figure out which server(s) to send the alterjob or runjob request to, so it'll be more performant.

File: src/lib/Libcmds/batch_status.c
Comment (suresh-thelkar) [reply src/lib/Libcmds/batch_status.c:139]: Renamed this as bs_init.

File: src/scheduler/data_types.h
Comment (suresh-thelkar) [reply src/scheduler/data_types.h:576]: I thought of this originally and did not add it for reservations also because they are shared across all the servers and we learn more when we implement them. Due to this I thought it is better add this to job_info object itself.

File: src/scheduler/pbs_sched.cpp
Comment (suresh-thelkar) [reply src/scheduler/pbs_sched.cpp:577]: Yes. Thanks for catching this. Please look into my latest commit for changes.

**Comment by suresh-thelkar** [comment]
> > Before you do a pbs_connect() and then use ANY IFL interface on that socket. Now that returns a multi-server socket that can only be used for certain calls. Now you need to use that multi-server interface to do a stat, and then find the socket you need to use for other calls like alterjob or runjob. Please write up a design for these changes so they can be discussed on the forums.
> 
> Suresh can explain more, but just to add some clarification, clients can still use the fd returned by pbs_connect directly without having to care about the fd inside each batch_status .. it's just that if they use the fd that comes within the batch_status, then IFL will not bother trying to figure out which server(s) to send the alterjob or runjob request to, so it'll be more performant.

Thanks Ravi. There is no change in using the IFLs as far as single server is concerned. It is because the virtual socket  returned by pbs_connect is actually real socket when it comes to single server configuration. 

When it comes to Multi-Server we use virtual socket to fetch node/job related stats. IFLs like pbs_sigjob, pbs_runjob from clients like scheduler are called on the real socket of the corresponding server for performance reasons.  I have created a design document and forum discussion for the same and tried to explain it in detail. Please look into that and provide your feedback.

Design Doc:
https://openpbs.atlassian.net/wiki/spaces/PD/pages/2233991174/Adding+a+new+member+to+batch+status+structure

Forum Discusssion:
https://community.openpbs.org/t/adding-a-new-member-to-batch-status-structure/2353

**Comment by suresh-thelkar** [comment]
@agrawalravi90 , I have implemented your review comments. Please check the latest commit.

@bhroam / @raviv1995 ,  Following are the links to design document and forum discussions. Please go through them and provide your feedback on the following also.
https://openpbs.atlassian.net/wiki/spaces/PD/pages/2233991174/Adding+a+new+member+to+batch+status+structure
https://community.openpbs.org/t/adding-a-new-member-to-batch-status-structure/2353

File: src/scheduler/data_types.h
Comment (agrawalravi90) [reply src/scheduler/data_types.h:576]: Well we haven't designed reservations yet, so I think we shouldn't assume that they will be shared, but at least while confirming a reservation, the sched does need to know which server sent it the reservation submit request so that it can respond it to that server, right?

File: openpbs.spec
Comment (agrawalravi90) [inline openpbs.spec:408]: why were these changes needed?

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]

File: openpbs.spec
Comment (suresh-thelkar) [reply openpbs.spec]: After dev install this file showed up in check-out list along with other files and checked it in without noticing it. We don't need this. I have removed it in the latest commit. Please check.

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]

File: openpbs.spec
Comment (hirenvadalia) [reply openpbs.spec:1]: Its used by Travis (to install build nd runtime dependency), its not safe to delete.

Comment (suresh-thelkar) [reply openpbs.spec:1]: I was thinking pbspro.spec.in is enough because psbpro.spec is generated from it automatically. It seems pbspro.spec is also used by Travis. So I keep both of them in my next commit.

**Comment by bhroam** [comment]
I've WIP'd this PR until the design discussion on the forum is complete (which it almost is)

File: src/scheduler/data_types.h
Comment (bhroam) [inline src/scheduler/data_types.h:742]: We might not need the node socket (and we don't need it right now).  We always talk to the job's server.  It might be true the job might move to another server, but the scheduler talks to the server which owns the job now.  We do need some way to match server to node, but that might be done via name and not socket.

Comment (suresh-thelkar) [reply src/scheduler/data_types.h:742]: Yes @bhroam. As per the updated design doc we don't need server_sock. We are actually getting server's fd by using server_instance_id which actually consists of "server_instance_name:port". I request you to go through the updated design document.

**Comment by suresh-thelkar** [comment]
> I've WIP'd this PR until the design discussion on the forum is complete (which it almost is)

I have now modified the design document to reflect the approach where we store server instance name as part of server_instance_id due to the reasons that I have mentioned in the forum. So yes I hope the design document is complete now unless there are any further questions. 

Marking this PR as WIP till I finish and test the code changes as per the updated design document.

File: src/scheduler/data_types.h
Comment (bhroam) [reply src/scheduler/data_types.h:742]: Ahh, the design has indeed updated since I last reviewed it.  I'd rename this from server_sock, to svr_inst_id since it is not a socket.

**Comment by suresh-thelkar** [comment]
@bhroam / @agrawalravi90,  As the design is signed off, and we are going with the new approach I have made the code changes accordingly. Had to squash and rebase to resolve conflicts and to make Travis work and run smoothly. It helps if you could review all the changes since beginning and provide your comments.

Also attached are the latest detailed test logs for Multi-Server.
[SchedJobRelatedAPIS.txt.log](https://github.com/openpbs/openpbs/files/5595994/SchedJobRelatedAPIS.txt.log)

**Comment by suresh-thelkar** [comment]
It seems Travis builds are not working at the moment due to some issue. I have run all the tests we usually run in Travis. All of them ran successfully.

File: src/include/svrfunc.h
Comment (agrawalravi90) [inline src/include/svrfunc.h:81]: void the argument

File: src/lib/Libattr/master_job_attr_def.xml
Comment (agrawalravi90) [inline src/lib/Libattr/master_job_attr_def.xml:2089]: Why 'ATR_DFLAG_NOSAVM' ? if you retrieve a job from db, won't this information be useful ?

File: src/scheduler/job_info.cpp
Comment (agrawalravi90) [inline src/scheduler/job_info.cpp:1240]: do a NULL check here as well?

File: src/server/process_request.c
Comment (agrawalravi90) [inline src/server/process_request.c:1678]: the svr_inst_name should match its PSI value. I think when you use gethostname(), you will get the same svr_inst_name for all servers running on that host, even though in PSI the the server name values might be different. So, I think we should use PBS_SERVER_NAME value instead of the hostname here, which is what PSI uses.

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]

File: src/scheduler/job_info.cpp
Comment (bhroam) [reply src/scheduler/job_info.cpp:1240]: No need to do a NULL check to just print a message.  That happens inside string_dup().  Although error recovery here is kind of important.

**Review by bhroam — CHANGES_REQUESTED** [review]

File: src/lib/Libattr/master_job_attr_def.xml
Comment (suresh-thelkar) [reply src/lib/Libattr/master_job_attr_def.xml:2089]: Yes, there is a possibility that the server instance name can be changed before it starts. This is the reason I have marked this attribute as ATR_DFLAG_NOSAVM. This is my understanding when I looked at the following code(pbsd_init_job function) with respect to the single server.

We are actually freeing the attribute JOB_ATR_at_server and every time populating its value. So Ideally this could have also been marked as ATR_DFLAG_NOSAVM.


```
	/* update at_server attribute in case name changed */
	free_jattr(pjob, JOB_ATR_at_server);
	set_jattr_generic(pjob, JOB_ATR_at_server, server_name, NULL, SET)
```
@subhasisb, please let us know your opinion on this.

File: src/server/process_request.c
Comment (suresh-thelkar) [reply src/server/process_request.c:1678]: When it comes to the context of Multi-Server, PBS_SERVER refers to cluster name. This may or may not be a hostname of one of the server instances. It could totally be different from a hostname. 

Yes the idea is svr_inst_name should match with one of the PSI values. We also compare port numbers when all the server instances run on a same machine. So even if gethostname()/get_fullhostname() are the same in the above case, we also compare its port numbers, so it is not a problem. Also, Admin when configuring PBS_SERVER_INSTANCES he should make sure that he always either uses short hostname or FQDN to refer to a server instance.  

In case if we are using PBS_SERVER while populating pbs_conf.psi while only port number is provided by Admin for a server instance, we should change that to actually use the respective server's instance name. We are not supposed to use PBS_SERVER name here.

@subhasisb, please correct me If I am wrong.

**Comment by hirenvadalia** [comment]
closing and reopening to restart travis.

File: src/server/process_request.c
Comment (agrawalravi90) [reply src/server/process_request.c:1678]: ok, can you please document this? That PSI should have valid hostnames, not PBS_SERVER (if it isn't documented already). Also, we should probably add a check inside pbs_loadconf() which verifies that each PSI value is a valid hostname.

Comment (agrawalravi90) [reply src/server/process_request.c:1678]: This also raises questions about how we'll handle talking to multiple servers in some of the IFL calls. Right now, we are in the process of adding multi-server support to IFLs like pbs_deljob, pbs_sigjob etc., which just take a jobid, which won't help identify the server if PSI will not be created using PBS_SERVER. So, if we do what you are proposing, then we'll need a different way to map jobids to servers.

Comment (agrawalravi90) [reply src/server/process_request.c:1678]: Actually, with this restriction, all servers will generate the same jobids, there will be job id conflicts, regardless of whether the servers run on the same host or on different hosts. I think your change cannot go in without fixing the jobid namespace conflict. Please let me know if I'm misunderstanding the implications of this.

Comment (suresh-thelkar) [reply src/server/process_request.c:1678]: > ok, can you please document this? That PSI should have valid hostnames, not PBS_SERVER (if it isn't documented already). Also, we should probably add a check inside pbs_loadconf() which verifies that each PSI value is a valid hostname.

Yes we have not documented the above when PBS_SERVER_INSTANCES are used. Yes  I am planning to do exactly same thing i.e. add code changes to make sure PSI is a valid hostname. Also planning to document it. Don't know why we missed it earlier.

Comment (suresh-thelkar) [reply src/server/process_request.c:1678]: > This also raises questions about how we'll handle talking to multiple servers in some of the IFL calls. Right now, we are in the process of adding multi-server support to IFLs like pbs_deljob, pbs_sigjob etc., which just take a jobid, which won't help identify the server if PSI will not be created using PBS_SERVER. So, if we do what you are proposing, then we'll need a different way to map jobids to servers.

Depending on jobid's hostname to figure out which is its server may or may not be the right idea. It is because going forward PBS_SERVER may or may not be hostname of one of the server instances configured. Instead, we can do the following.
An IFL API if it wants to find out which server is responsible for a given jobid, it tries to contact server instances in some random fashion and finds the right server. If the request reaches to a server which does not own the job it rejects with "unknown job id code". One of the server instances respond and process the request successfully. This way that IFL identifies which server it has to talk to. This is what @nithinj  is doing as part of his PR https://github.com/openpbs/openpbs/pull/2086

Comment (suresh-thelkar) [reply src/server/process_request.c:1678]: > Actually, with this restriction, all servers will generate the same jobids, there will be job id conflicts, regardless of whether the servers run on the same host or on different hosts. I think your change cannot go in without fixing the jobid namespace conflict. Please let me know if I'm misunderstanding the implications of this.

The restriction to have same value for PBS_SERVER is not enforced in the code as of now. We need to keep this assumption in our minds and code it accordingly. So we are not using its(PBS_SERVER) value while configuring server instances and the code in this PR is exactly doing that. So given below is a mechanism how we can test as of now. 

If we make servers run on different complexes then we will get jobids differently because PBS_SERVER value is different. For example, we can get jobids like **1.spark,1.spark2, 2.spark, 2.spark2, 3.spark, 3.spark2**
In one of my test runs(as part of my PR testing) I have got the following
3.spark         root     workq    STDIN       50965   1   1    --    --  R 00:11
    --
3.spark2        root     workq    STDIN       73131   1   1    --    --  R 00:00
   Job run at Wed Nov 25 at 08:51 on (spark2:ncpus=1)
 
 
Also, when all server instances run on the same host, I think we can make aliases for PBS_SERVER in /etc/hosts and use it I believe,  in which case we get the jobids similar to the above.
 
Once we solve "jobid namespace problem" then we can strictly use only one value for PBS_SERVER across all complexes or across all server instances. In fact, we can enforce this in the code. When we do this we don't need to change any of the current code that is added as part of this PR.  This way I believe we don't need to wait till the completion of "jobid namespace feature".

**Comment by suresh-thelkar** [comment]
@bhroam / @agrawalravi90, I have addressed all your comments. Please take a look.

Also, attached are Multi-Server testing logs after implementing the review comments. All test cases passed.
[SchedJobRelatedAPIS_Latest.txt](https://github.com/openpbs/openpbs/files/5607549/SchedJobRelatedAPIS_Latest.txt)

**Comment by suresh-thelkar** [comment]
Attached are TH3 regression logs after implementing the review comments.

Description: Rerun All Tests From #5022
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|5059|3843|1|0|0|0|3842|

It seems one test case that failed is happening in master also.

**Comment by suresh-thelkar** [comment]
@bhroam / @agrawalravi90, I have also added the new attribute the pbs_ifl_mock.py which is needed for few PTL tests. Please take a look at this commit as well.

**Comment by suresh-thelkar** [comment]
Latest full regression logs as of today.
Description: Rerun All Tests From #5073
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|5076|3843|3|0|0|0|3840|

Test cases failed are also observed in nightly.

File: src/scheduler/job_info.cpp
Comment (agrawalravi90) [inline src/scheduler/job_info.cpp:128]: why did you include this file?

File: src/server/process_request.c
Comment (agrawalravi90) [reply src/server/process_request.c:1674]: > An IFL API if it wants to find out which server is responsible for a given jobid, it tries to contact server instances in some random fashion and finds the right server. If the request reaches to a server which does not own the job it rejects with "unknown job id error code". One of the server instances(which is the owing server instance for the job) respond and process the request successfully. This way the IFL identifies which server it has to talk to. This is what @nithinj is doing as part of his PR #2086

Well that's not ideal for performance (both client and server side). Think about qdel list which was added recently .. we can't just broadcast the entire list of jobs to all servers, that will slow both client and server sides.

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]

File: src/scheduler/job_info.cpp
Comment (suresh-thelkar) [reply src/scheduler/job_info.cpp]: As per the latest changes this header file libpbs.h inclusion is not needed. I have removed it in the latest commit. Please check.

File: src/server/process_request.c
Comment (suresh-thelkar) [reply src/server/process_request.c:1674]: The approach that I mentioned is being followed in the PR #2086 for most of the IFL APIs in general when they are called from a non scheduler like client/s. You can go through that PR for more details. 

When it comes to IFLs like pbs_deljob where performance matters we cannot follow the same approach. Instead, we can go through the following method to achieve the desired goal. There might be other better ways. Following are just my thoughts.
- Do a stat of the required jobs inside the IFL __pbs_deljoblist. Group the jobs by server_instance_id.
- Iterate over all the resultant server instances and pass one job-list at a time to the function PBSD_deljoblist().
- As the number of server instances configured are minimum this greatly enhances the performance.

Also, we are already following a slightly different approach for some IFLs like pbs_runjob for example when qrun with '-H' option is called. We actually do a stat for the given node through which we can find the right server instance id and send the request to that server instance directly instead of contacting servers in some random fashion and finding out which is the right server instance.

**Comment by suresh-thelkar** [comment]
Given below are very lates regression results. 

Description: Tests from given sources on platforms UBUNTU1804, SUSE15, UBUNTU1604, CENTOS8, RHEL8, CENTOS7, RHEL7, SLES12, SLES15, WIN2K16
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|5082|3843|1|0|0|0|3842|

One test case that failed is also observed in nightly.

**Comment by suresh-thelkar** [comment]
@agrawalravi90, I have addressed your latest review comments also. Please take a look.

File: src/scheduler/fifo.cpp
Comment (bhroam) [inline src/scheduler/fifo.cpp:1649]: Please have a shortcut in get_svr_inst_fd() if there is only 1 server.  No need to search by string if it is just one.

File: src/scheduler/job_info.cpp
Comment (bhroam) [inline src/scheduler/job_info.cpp]: Don't you want to send the alter to the specific server?

**Review by bhroam — CHANGES_REQUESTED** [review]

File: src/server/process_request.c
Comment (bhroam) [reply src/server/process_request.c:1674]: Wait, I thought a runjob always went to the server currently owning the job.  The owning server is responsible to move and run the job.

Comment (agrawalravi90) [reply src/server/process_request.c:1674]: ya, we always send it to the job owner, I think Suresh is mistaken, but @nithinj can confirm once.

File: src/scheduler/fifo.cpp
Comment (suresh-thelkar) [reply src/scheduler/fifo.cpp:1649]: Optimized get_svr_inst_fd() to return early in case of single server mode. Please check the latest commit. 

Please note that in case of single server we cannot return vfd itself. It is because it may not be the first fd of the server instances after the PR #2086.

File: src/scheduler/job_info.cpp
Comment (suresh-thelkar) [reply src/scheduler/job_info.cpp]: Yes, But I am not modifying pbs_asyalterjob as part of this PR as we have to handle this differently. I added a note about the same in the conversations section(last bullet point). Please check. We can take this after this PR.

**Comment by suresh-thelkar** [comment]
@bhroam, I have addressed your comments. Please take a look.

File: src/scheduler/fifo.cpp
Comment (nithinj) [reply src/scheduler/fifo.cpp:1649]: just to clarify, vfd and real fd will be the same in the case of a single server cluster. Returning the first fd also works.

Comment (agrawalravi90) [reply src/scheduler/fifo.cpp:1649]: that's how it is right now, but aren't we planning to change it so that vfd is not the first real fd? To be able to distinguish between real and virtual fd in IFL ?

File: src/scheduler/job_info.cpp
Comment (agrawalravi90) [reply src/scheduler/job_info.cpp]: you don't have to modify pbs_asyalterjob Suresh if you just send it the right fd, right?

Comment (agrawalravi90) [reply src/scheduler/job_info.cpp]: without this, i don't see how this code will work for multi-server, it'll always send alterjob to the first server if we sent it the virtual fd

**Review by bhroam — APPROVED** [review]
Looks good.  Please fix the conflict.

File: src/scheduler/fifo.cpp
Comment (nithinj) [reply src/scheduler/fifo.cpp:1649]: I was talking about a single server cluster. For multi-server cluster vfd will be different than that of any instance fd.

File: src/scheduler/job_info.cpp
Comment (nithinj) [reply src/scheduler/job_info.cpp]: We can take it up as part of #2086 once this gets merged.

File: src/scheduler/fifo.cpp
Comment (agrawalravi90) [reply src/scheduler/fifo.cpp:1649]: ah ok, makes sense

**Comment by agrawalravi90** [comment]
> Looks good. Please fix the conflict.

Since Suresh is out on vacation and this is kind of urgent, and the conflict has to do with my master branch changes, I'll take it up

**Comment by bhroam** [comment]
@agrawalravi90 do you sign off on the PR?  You are the other reviewer.  If you are happy with it, you should sign off before you work on it.

**Review by agrawalravi90 — APPROVED** [review]

File: src/server/process_request.c
Comment (nithinj) [reply src/server/process_request.c:1674]: sorry, I missed this thread earlier.
qrun sends the request to the job owning server. IFL will mention where the mother superior resides by doing a node_stat. It is a special case as we are dealing with two shared objects instead of one.

**Pull request merged**
